x/valset-pref/keeper.go

```diff
@@ -49,42 +50,63 @@ func (k Keeper) GetDelegationPreferences(ctx sdk.Context, delegator string) (typ
-func (k Keeper) GetExistingStakingDelegations(ctx sdk.Context, delAddr sdk.AccAddress) ([]types.ValidatorPreference, error) {
-	var existingDelsValSetFormatted []types.ValidatorPreference
+// GetValSetPreferencesWithDelegations fetches the delegator's validator set preferences
+// considering their existing delegations.
+// -If validator set preference does not exist and there are no existing delegations, it returns an error.
+// -If validator set preference exists and there are no existing delegations, it returns the existing preference.
+// -If there is any existing delegation:
+// calculates the delegator's shares in each delegation
+// as a ratio of the total shares and returns it as part of ValidatorSetPreferences.
+func (k Keeper) GetValSetPreferencesWithDelegations(ctx sdk.Context, delegator string) (types.ValidatorSetPreferences, error) {
+	delAddr, err := sdk.AccAddressFromBech32(delegator)
+	if err != nil {
+		return types.ValidatorSetPreferences{}, err
+	}
 
+	valSet, exists := k.GetValidatorSetPreference(ctx, delegator)
 	existingDelegations := k.stakingKeeper.GetDelegatorDelegations(ctx, delAddr, math.MaxUint16)
-	if len(existingDelegations) == 0 {
-		return nil, types.ErrNoDelegation
+
+	// No existing delegations for a delegator when valSet does not exist
+	if !exists && len(existingDelegations) == 0 {
+		return types.ValidatorSetPreferences{}, fmt.Errorf("No Existing delegation to unbond from")
 	}
 
-	existingTotalShares := osmomath.NewDec(0)
-	// calculate total shares that currently exists
-	for _, existingDelegation := range existingDelegations {
-		existingTotalShares = existingTotalShares.Add(existingDelegation.Shares)
+	// Returning existing valSet when there are no existing delegations
+	if exists && len(existingDelegations) == 0 {
+		return valSet, nil
 	}
 
-	// for each delegation format it in types.ValidatorSetPreferences format
-	for _, existingDelegation := range existingDelegations {
-		existingDelsValSetFormatted = append(existingDelsValSetFormatted, types.ValidatorPreference{
-			ValOperAddress: existingDelegation.ValidatorAddress,
-			Weight:         existingDelegation.Shares.Quo(existingTotalShares),
-		})
+	// when existing delegation exists, have it based upon the existing delegation
+	// regardless of the delegator having valset pref or not
+	return types.ValidatorSetPreferences{Preferences: formatToValPrefArr(existingDelegations)}, nil
+}
+
+// formatToValPrefArr iterates over given delegations array, formats it into ValidatorPreference array.
+// Used to calculate weights for the each delegation towards validator.
+// CONTRACT: This method assumes no duplicated ValOperAddress exists in the given delegation.
+func formatToValPrefArr(delegations []stakingtypes.Delegation) []types.ValidatorPreference {
+	totalShares := sdk.NewDec(0)
+	for _, existingDelegation := range delegations {
+		totalShares = totalShares.Add(existingDelegation.Shares)
```

Comment:
Wait... is this the code for valset preference? I think this is fundamentally wrong. This needs to be based on the token amount, not the number of shares